pkg/resource/detector/detector.go

```diff
@@ -0,0 +1,101 @@
+package detector
+
+import (
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"time"
+)
+
+// Detector represents a procedure that runs in the background, periodically auto-detecting features
+type Detector struct {
+	dc                  discovery.DiscoveryInterface
+	ticker              *time.Ticker
+	SubscriptionChannel chan schema.GroupVersionKind
+	crds 				map[runtime.Object]trigger
+}
+
+type trigger func(runtime.Object)
+
+// New creates a new auto-detect runner
+func NewAutoDetect(dc discovery.DiscoveryInterface) (*Detector, error) {
+	// Create a new channel that GVK type will be sent down
+	subChan := make(chan schema.GroupVersionKind, 1)
+
+	return &Detector{dc: dc, SubscriptionChannel: subChan, crds:map[runtime.Object]trigger{}}, nil
+}
+
+
+//AddCRDTrigger to run the trigger function,
+//the first time that the background scanner discovers that the CRD type specified exists
+func (d *Detector) AddCRDTrigger(crd runtime.Object, trigger trigger) {
+	d.crds[crd] = trigger
+}
+
+//AddCRDsTrigger to run the trigger function,
+//the first time that the background scanner discovers that each of the CRD types specified exists
+func (d *Detector) AddCRDsTrigger(crds []runtime.Object, trigger trigger) {
+	for _, crd := range crds {
+		d.AddCRDTrigger(crd, trigger)
+	}
+}
+
+//AddCRDsWithTriggers to run the associated trigger function for the particular CRD,
+//the first time that the background scanner discovers that the CRD type specified exists
+func (d *Detector) AddCRDsWithTriggers(crdsTriggers map[runtime.Object]trigger) {
+	for crd, trigger := range crdsTriggers {
+		d.AddCRDTrigger(crd, trigger)
+	}
+}
+
+// Start initializes the auto-detection process that runs in the background
+func (d *Detector) Start(interval time.Duration) {
+	// periodically attempts to auto detect all the capabilities for this operator
+	d.ticker = time.NewTicker(interval * time.Second)
+
+	go func() {
+		d.autoDetectCapabilities()
+
+		for range d.ticker.C {
+			d.autoDetectCapabilities()
+		}
+	}()
+}
+
+// Stop causes the background process to stop auto detecting capabilities
+func (d *Detector) Stop() {
+	d.ticker.Stop()
+	close(d.SubscriptionChannel)
+}
+
+func (d *Detector) autoDetectCapabilities() {
+	for crd, trigger := range d.crds {
+		crdGVK := crd.GetObjectKind().GroupVersionKind()
+		resourceExists, _ := d.resourceExists(d.dc, crdGVK.GroupVersion().String(), crdGVK.Kind)
+		if resourceExists {
+			stateManager := GetStateManager()
+			if stateManager.GetState(crdGVK.Kind) != true {
+				stateManager.SetState(crdGVK.Kind, true)
+				trigger(crd)
+			}
+		}
+	}
+}
+
+//copied from operator-sdk to avoid pulling in thousands of files in imports
```

Comment:
I'd suggest putting a link (permalink, using commit SHA rather than e.g. `master`) here in this comment.

Alternatively, since this project is using `dep` rather than Go modules, you could add the operator-sdk as a dependency in Gopkg.toml & set the prune options for it to prune unused packages, so that only the packages that you've explicitly imported in the code would ever be included.

https://github.com/golang/dep/blob/master/docs/Gopkg.toml.md#prune

I'm not familiar enough with Go modules to know if there's an equivalent in that.